File: 2025_05/leetcode2025-05-28.py
from typing import List
from collections import deque

class Solution:
    def maxTargetNodes(self, edges1: List[List[int]], edges2: List[List[int]], k: int) -> List[int]:
        if k == 0:
            return [1]*(len(edges1)+1)
        
        # 首先根据edges1和edges2分别建出第一个图和第二个图g1和g2，接下来用bfs的方法，在g1中从每一个节点出发，一层一层地循环，一共循环k层，在循环的同时进行统计，这样我们就得到了在g1中从每一个节点出发路径数量小于等于k的目标节点，将它们按顺序存入列表d1中
        # 同时我们还需要计算g2，但是需要注意的是，如果用一条边连接了两个图，那么对于g1中的每一个点，如果要在g2中找路径数量小于等于k的节点，就需要从g2中被连接的那个节点出发，寻找在g2中路径数量小于k的节点数量。那么与上面的bfs同理，只需将路径长度限制改为k-1，同时更改图为g2即可
        # 于是我们可以编写一个函数generate_dis，传入一个图和一个路径长度限制，用来计算该图中从各个节点出发有多少个节点满足到这个节点的路径长度小于等于给出的限制的，同时我们还有一个bfs函数，传入一个图和一个点以及路径长度限制，返回从当前点开始，满足与当前点的距离小于等于限制的点的个数，也就是说，generate_dis函数只需对每个节点调用一次bfs函数即可

        def generate_dis(edges, k):
            n = len(edges)+1
            g = [[] for _ in range(n)]
            for a,b in edges:
                g[a].append(b)
                g[b].append(a)

            res = []
            for i in range(n):
                # 对每个点计算一次
                res.append(bfs(g, i, k))
            return res

        def bfs(g, i, k):
            n = len(g)
            q = deque()
            q.append(i)
            visited = [0]*n # 防止重复计算
            ans = 0
            depth = 0 # 防止路径超出限制
            while q and depth<k:
                depth += 1
                cur = []
                for x in q:
                    # 每次更新为下一层
                    visited[x] = 1
                    for y in g[x]:
                        if visited[y]:
                            continue
                        cur.append(y)
                        ans += 1
                q = cur
            return ans

        d1, d2 = generate_dis(edges1, k), generate_dis(edges2, k-1) # 注意第二张图只需要k-1的长度
        tree2max = max(d2)+1 # 并且我们只需要取第二张图满足条件的最大值即可，每次连接只连接只一个点
        res = [0]*(len(edges1)+1)
        for i in range(len(edges1)+1):
            res[i] = d1[i]+tree2max+1
        return res

if __name__ == "__main__":
    s = Solution()
    edges1 = [[0,1],[0,2],[2,3],[2,4]]
    edges2 = [[0,1],[0,2],[0,3],[2,7],[1,4],[4,5],[4,6]]
    k = 2
    print(s.maxTargetNodes(edges1,edges2,k))

File: 2025_05/leetcode2025-05-28.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
private:
    vector<vector<int>> build_tree(vector<vector<int>>& edges){ // 建图函数
        int n = edges.size()+1;
        vector<vector<int>> g(n);
        for (auto& e : edges){
            int a = e[0], b = e[1];
            g[a].push_back(b);
            g[b].push_back(a);
        }
        return g;
    }

    int dfs(int x, int pa, int depth, int K, vector<vector<int>>& g){
        if (depth > K){
            // 保证深度不超过路径长度限制
            return 0;
        }
        int ans = 1; // 保存当前节点x
        for (int y : g[x]){
            if (y == pa){
                continue;
            }
            ans += dfs(y, x, depth+1, K, g);
        }
        return ans;
    }

public:
    vector<int> maxTargetNodes(vector<vector<int>>& edges1, vector<vector<int>>& edges2, int k) {
        // 也可以使用dfs的方法，每次传入一个图，给出一个节点，那么直接用深度优先搜索找出所有深度小于等于给出的路径长度限制的节点个数即可。即dfs(x, pa, depth, k, g)的返回值就是在g中从x出发所有深度小于等于k的节点个数
        // 首先对于edges2中的每一个点我们都计算一次，求出从第二个图中的某个节点出发，深度小于等于k-1的最大节点个数，那么对于edges1中的每一个点只需计算出深度小于等于k的最大节点个数，加上edges2中计算出的最大值，即可得到答案
        if (k == 0){
            return vector<int>(edges1.size()+1, 1);
        }
        int n = edges1.size()+1, m = edges2.size()+1;
        vector<vector<int>> g1 = build_tree(edges1), g2 = build_tree(edges2);
        int tree2max = 0; // 记录edges2中的最大合法节点值
        for (int i=0; i<m; i++){
            tree2max = max(tree2max, dfs(i, -1, 0, k-1, g2));
        }

        vector<int> res(n, 1);
        for (int i=0; i<n; i++){
            // 因为已经有edges2中最大合法节点值，那每一次都将那个节点连接到edges1中的第i个节点即可
            // 所以只需计算出edges1中从i出发深度小于等于k的节点数量相加即可
            res[i] = dfs(i, -1, 0, k, g1)+tree2max;
        }
        return res;
    }
};

// class Solution {
// private:
//     vector<int> generate_dis(vector<vector<int>>& edges, int k){
//         int n = edges.size()+1;
//         vector<vector<int>> g(n);
//         for (auto& e : edges){
//             int a = e[0], b = e[1];
//             g[a].push_back(b);
//             g[b].push_back(a);
//         }

//         vector<int> res(n);
//         for (int i=0; i<n; i++){
//             res[i] = bfs(g, i, k);
//         }
//         return res;
//     }

//     int bfs(vector<vector<int>>& g, int i, int k){
//         int n = g.size();
//         queue<int> q;
//         q.push(i);
//         int depth = 0;
//         int ans = 0;
//         vector<int> visited(n, 0);
//         while (q.size() && depth<k){
//             depth += 1;
//             queue<int> cur;
//             while (!q.empty()){
//                 int x = q.front();
//                 q.pop();
//                 visited[x] = 1;
//                 for (int y : g[x]){
//                     if (visited[y]){
//                         continue;
//                     }
//                     cur.push(y);
//                     ans += 1;
//                 }
//             }
//             q = move(cur);
//         }
//         return ans;
//     }

// public:
//     vector<int> maxTargetNodes(vector<vector<int>>& edges1, vector<vector<int>>& edges2, int k) {
//         int n = edges1.size()+1;
//         if (k == 0){
//             return vector<int>(n, 1);
//         }

//         vector<int> d1 = generate_dis(edges1, k), d2 = generate_dis(edges2, k-1);
//         int tree2max = 0;
//         for (int i : d2){
//             tree2max = max(tree2max, i);
//         }
//         tree2max++;
//         vector<int> res(n);
//         for (int i=0; i<n; i++){
//             res[i] = d1[i]+1+tree2max;
//         }
//         return res;
//     }
// };

int main(){
    Solution s;
    vector<vector<int>> edges1 = {{0,1},{0,2},{2,3},{2,4}}, edges2 = {{0,1},{0,2},{0,3},{2,7},{1,4},{4,5},{4,6}};
    int k = 2;
    vector<int> res = s.maxTargetNodes(edges1, edges2, k);
    for (int i : res){
        cout << i << " ";
    }
    cout << endl;
}
